### PR TITLE
Master pricelist pocalypse vfe

### PR DESCRIPTION
**Pull request by Feyensv**
### WIP

- [ ] Technical Cleanup
  - [x] Product
  - [x] Sale_*
  - [x] Website_sale_*
- [x] Pricelist no longer required to make Odoo flows work (sale, subscription, point_of_sale, website_sale, repair)
  - [x] Currency of Orders = pricelist currency / company currency
  - [x] When no Pricelist defined, fallback on list price (converted to given currency if needed).
- [ ] Technical Cleanup :
  * Multi - Uom, currencies, companies cleanup and fixes
  * Performance improvements ?
- [x] Pricelist application on top of event/rental/subscription prices computation
- [x] Consistent discount computation between frontend and backend (sale/website_sale)
- [ ] Do not base tests/flows/behaviors on the existence of the Public Pricelist.
  - [ ] Tests
  - [ ] Pricelist creation on company currency changes
  - [x] Flows

#### Related Tasks
- [ ] Pricelist Report
- [ ] On pricelist change, ability to recompute all prices

...

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by Feyensv** [comment]
tuple for ctxt key ptav_ids
Fix 13.0 pricelists ctxt key for compute fct depending on search.

File: addons/point_of_sale/static/src/js/models.js [hunk 1]
```diff
@@ -395,14 +396,21 @@ exports.PosModel = Backbone.Model.extend({
...
+            };
+        },
+        loaded: function (self, products) {
             var using_company_currency = self.config.currency_id[0] === self.company.currency_id[0];
             var conversion_rate = self.currency.rate / self.company_currency.rate;
             self.db.add_products(_.map(products, function (product) {
                 if (!using_company_currency) {
-                    product.lst_price = round_pr(product.lst_price * conversion_rate, self.currency.rounding);
+                    // The Sales price is already converted by the currency_id in the context.
+                    product.standard_price = round_pr(product.lst_price * conversion_rate, self.currency.rounding);
```
Comment (Feyensv) [inline addons/point_of_sale/static/src/js/models.js]: product.standard_price in the conversion...

**Comment by nhomar** [comment]
Friend.

can you paste in the description the pad with the specifications?

THANKS A LOT.

**Comment by Feyensv** [comment]
I'm updating the PR header on the fly with them (the pad isn't really clear for the moment), but even the specs can be considered to be in draft :D.
That's why it's just a draft PR, don't take all changes here for what's going to really change.
For the moment, nearly no functional changes have been decided, I'm mainly cleaning the technical mess, making sure currencies, uom, and multi company works fine.

The only two confirmed functional objectives for the moment are :

* being able to do the Sales flow even if all pricelists are removed (fallback to Sales price).
* Apply pricelists on top of custom pricings (event, subscription, rental) to allow discounts/formulas for those flows.

An internal RFC is on the way to consider what's missing/not intuitive in the pricelist currently.  And I'm still thinking about making a RFC to the community on github ;).

File: addons/event_sale/models/sale_order.py [hunk 2]
```diff
@@ -70,21 +89,22 @@ def _onchange_event_id(self):
...
-    @api.onchange('product_uom', 'product_uom_qty')
-    def product_uom_change(self):
-        if not self.event_ticket_id:
-            super(SaleOrderLine, self).product_uom_change()
+    def _compute_price(self):
+        self.ensure_one()
+        if self.event_ticket_id and self.event_id:
+            company = self.company_id or self.event_id.company_id or self.env.company
+            currency = company.currency_id
+            # VFE FIXME what is the currency of the price on the event ticket ???
```
Comment (Feyensv) [inline addons/event_sale/models/sale_order.py]: use currency of event_ticket

File: addons/point_of_sale/models/pos_order.py [hunk 3]
```diff
@@ -788,12 +788,13 @@ def _compute_amount_line_all(self):
     @api.onchange('product_id')
     def _onchange_product_id(self):
         if self.product_id:
+            # VFE TODO support no pricelist in pos also...
```
Comment (Feyensv) [inline addons/point_of_sale/models/pos_order.py]: Done, remove

File: addons/product/models/product.py [hunk 4]
```diff
@@ -215,47 +215,13 @@ def _compute_combination_indices(self):
...
-            if self._context.get('uom'):
-                value = self.env['uom.uom'].browse(self._context['uom'])._compute_price(product.price, product.uom_id)
-            else:
-                value = product.price
-            value -= product.price_extra
-            product.write({'list_price': value})
+        self.is_product_variant = True
 
     def _set_product_lst_price(self):
+        # VFE TODO raise Error if multiple variants on template of the product ???
```
Comment (Feyensv) [inline addons/product/models/product.py]: (NB: the field is readonly when product_variant_count > 0, thus writing on it in this case is already error-prone).

File: addons/product/models/product_attribute.py [hunk 5]
```diff
@@ -388,11 +388,11 @@ class ProductTemplateAttributeValue(models.Model):
     attribute_line_id = fields.Many2one('product.template.attribute.line', required=True, ondelete='cascade', index=True)
 
     # configuration fields: the price_extra and the exclusion rules
-    price_extra = fields.Float(
-        string="Value Price Extra",
-        default=0.0,
+    price_extra = fields.Monetary(
+        string="Value Price Extra", # VFE TODO digits still necessary ???
```
Comment (Feyensv) [inline addons/product/models/product_attribute.py]: remove digits

File: addons/product/models/product_pricelist.py [hunk 6]
```diff
@@ -212,99 +205,154 @@ def _compute_price_rule(self, products_qty_partner, date=False, uom_id=False):
...
+            date=date,
+        ).get(product.id, [product.price_compute(
+            'list_price',
+            uom=uom,
+            date=date,
+            currency=currency
+            )[product.id]]
+        )[0]
+
+    # VFE TODO get_product_price_rule's' renaming
```
Comment (Feyensv) [inline addons/product/models/product_pricelist.py]: remove

File: addons/product/models/product_template.py [hunk 7]
```diff
@@ -10,6 +10,83 @@
...
+        fixed_price_currency = self.env['res.currency'].browse(self.env.context.get('fixed_sales_currency'))
+
+    ptavs = self.env['product.template.attribute.value'].browse(self.env.context.get('ptav_ids', []))
+    targeted_product_id = self.env.context.get('p_id', len(self) == 1)
+    if ptavs:
+        if self[0].is_product_variant:
+            # If we are computing the price on variants,
+            # We shouldn't add extras from ptavs which should be on the variant already.
+            ptavs = ptavs - self[0].product_template_attribute_value_ids
+        # VFE TODO filter for dynamic attributes if on current product...
```
Comment (Feyensv) [inline addons/product/models/product_template.py]: remove

File: addons/product/models/product_template.py [hunk 8]
```diff
@@ -71,29 +149,30 @@ def _read_group_categ_id(self, categories, domain, order):
...
-    # price fields
-    # price: total template price, context dependent (partner, pricelist, quantity)
-    price = fields.Float(
-        'Price', compute='_compute_template_price', inverse='_set_template_price',
-        digits='Product Price')
     # list_price: catalog price, user defined
-    list_price = fields.Float(
+    list_price = fields.Monetary(
         'Sales Price', default=1.0,
         digits='Product Price',
```
Comment (Feyensv) [inline addons/product/models/product_template.py]: remove all digits

File: addons/product/models/product_template.py [hunk 9]
```diff
@@ -183,47 +262,56 @@ def _compute_currency_id(self):
...
+            self.env.context.get('date', fields.Date.today()),
+            target_currency,
+        )
+
+    @api.depends_context('pricelist_id', 'quantity', 'uom_id', 'date', 'ptav_ids', 'currency_id')
+    def _compute_price(self):
+        # Computed for templates AND variants
+        # VFE TODO support currency_id in context like for product_product compute_price ?
+        # VFE TODO display_currency_id for the context dependent price ?????????
+        pricelist, quantity, uom, date = self._get_context_values()
```
Comment (Feyensv) [inline addons/product/models/product_template.py]: pricelist, quantity, uom, date, currency !

File: addons/product/models/res_partner.py [hunk 10]
```diff
@@ -11,7 +11,7 @@ class Partner(models.Model):
     # NOT A REAL PROPERTY !!!!
     property_product_pricelist = fields.Many2one(
         'product.pricelist', 'Pricelist', compute='_compute_product_pricelist',
```
Comment (Feyensv) [inline addons/product/models/res_partner.py]: depends_context(company/force_company?)

File: addons/sale/models/sale.py [hunk 11]
```diff
@@ -109,8 +109,7 @@ def _compute_tax_id(self):
         """
         Trigger the recompute of the taxes if the fiscal position is changed on the SO.
         """
-        for order in self:
-            order.order_line._compute_tax_id()
+        self.order_line._compute_tax_id()
```
Comment (Feyensv) [inline addons/sale/models/sale.py]: No price update on taxes changes ? (for included taxes ?)

File: addons/sale/models/sale.py [hunk 12]
```diff
@@ -342,25 +355,28 @@ def onchange_partner_id(self):
...
-        addr = self.partner_id.address_get(['delivery', 'invoice'])
-        partner_user = self.partner_id.user_id or self.partner_id.commercial_partner_id.user_id
+        partner = self.partner_id
+        self = self.with_context(force_company=(self.company_id or self.env.company).id).with_context(lang=partner.lang)
+        addr = partner.address_get(['delivery', 'invoice'])
+        partner_user = partner.user_id or partner.commercial_partner_id.user_id
         values = {
-            'pricelist_id': self.partner_id.property_product_pricelist and self.partner_id.property_product_pricelist.id or False,
-            'payment_term_id': self.partner_id.property_payment_term_id and self.partner_id.property_payment_term_id.id or False,
+            # VFE TODO don't set property_product_pricelist on all partners.
```
Comment (Feyensv) [inline addons/sale/models/sale.py]: Remove
* do we still want to set the pricelist magically if not given to create?

File: addons/sale/models/sale.py [hunk 13]
```diff
@@ -1065,13 +1088,15 @@ def _get_price_reduce_notax(self):
...
         for line in self:
+            # TODO update to support no partner set ?
             fpos = line.order_id.fiscal_position_id or line.order_id.partner_id.property_account_position_id
             # If company_id is set, always filter taxes by the company
             taxes = line.product_id.taxes_id.filtered(lambda r: not line.company_id or r.company_id == line.company_id)
             line.tax_id = fpos.map_tax(taxes, line.product_id, line.order_id.partner_shipping_id) if fpos else taxes
 
     @api.model
     def _get_purchase_price(self, pricelist, product, product_uom, date):
+        # VFE TODO ensure extensions support empty pricelist recordset.
```
Comment (Feyensv) [inline addons/sale/models/sale.py]: todo

File: addons/sale/models/sale.py [hunk 14]
```diff
@@ -1450,27 +1440,57 @@ def product_id_change(self):
...
-            date=self.order_id.date_order,
-            pricelist=self.order_id.pricelist_id.id,
-            uom=self.product_uom.id
+    def _update_description(self):
+        self.name = self.get_sale_order_line_multiline_description_sale(self.product_id)
+
+    def _compute_price(self):
+        self.ensure_one()
+
+        # VFE TODO POC remove no_variant attributes ...
```
Comment (Feyensv) [inline addons/sale/models/sale.py]: remove

File: addons/sale_coupon/models/sale_order.py [hunk 15]
```diff
@@ -425,11 +420,18 @@ def _compute_tax_id(self):
         # as the product is the generic discount line.
         # In case of a free product, retrieving the tax on the line instead of the product won't affect the behavior.
         for line in reward_lines:
+            # VFE FIXME correct company for fiscal position fetch
```
Comment (Feyensv) [inline addons/sale_coupon/models/sale_order.py]: use get_fiscal_position ?

File: addons/website_sale/models/sale_order.py [hunk 16]
```diff
@@ -69,83 +69,73 @@ def _search_abandoned_cart(self, operator, value):
...
+            [int(ptav['custom_product_template_attribute_value_id']) for ptav in custom_values])
+
+        if any(not ptav.is_custom for ptav in custom_ptavs):
+            # VFE FIXME consistency check, may be removed before merge.
+            raise UserError(_("An attribute has been given as custom while being not custom."))
+
+        # save is_custom attributes values
+        if custom_values:
+            # VFE TODO add missing custom ptav, with empty value.
+            # VFE TODO make it returned as value[ptav_id] = value
```
Comment (Feyensv) [inline addons/website_sale/models/sale_order.py]: needed ?

File: addons/sale/models/sale.py (hunk 12 above)
Comment (Feyensv) [reply addons/sale/models/sale.py]: Yes, to bring back

Comment (Feyensv) [reply addons/sale/models/sale.py]: Yes

File: addons/event_sale/models/event.py [hunk 17]
```diff
@@ -93,6 +97,11 @@ def _default_product_id(self):
     seats_unconfirmed = fields.Integer(string='Unconfirmed Seat Reservations', compute='_compute_seats', store=True)
     seats_used = fields.Integer(compute='_compute_seats', store=True)
 
+    @api.depends('event_id.company_id')
```
Comment (Feyensv) [inline addons/event_sale/models/event.py]: event_id/event_id.company_id/company_id ?

Comment (Feyensv) [reply addons/event_sale/models/event.py]: company_id

File: addons/sale/models/sale.py [hunk 18]
```diff
@@ -407,22 +429,21 @@ def _onchange_commitment_date(self):
...
+            vals['name'] = self.env['ir.sequence'].with_context(force_company=company_id).next_by_code(
+                'sale.order', sequence_date=seq_date) or _('New')
 
         # Makes sure partner_invoice_id', 'partner_shipping_id' and 'pricelist_id' are defined
         if any(f not in vals for f in ['partner_invoice_id', 'partner_shipping_id', 'pricelist_id']):
             partner = self.env['res.partner'].browse(vals.get('partner_id'))
             addr = partner.address_get(['delivery', 'invoice'])
             vals['partner_invoice_id'] = vals.setdefault('partner_invoice_id', addr['invoice'])
             vals['partner_shipping_id'] = vals.setdefault('partner_shipping_id', addr['delivery'])
+            # VFE TODO check if property_product_pricelist returns empty recordset if not set or not.
```
Comment (Feyensv) [inline addons/sale/models/sale.py]: Remove default pricelist assignation in separate commit.

File: addons/sale/models/sale.py [hunk 19]
```diff
@@ -1510,21 +1526,14 @@ def product_id_change(self):
         return result
 
     @api.onchange('product_uom', 'product_uom_qty')
-    def product_uom_change(self):
+    def quantity_change(self):
+        # VFE TODO remove the context brolls on non relational fields in view.
         if not self.product_uom or not self.product_id:
             self.price_unit = 0.0
             return
```
Comment (Feyensv) [inline addons/sale/models/sale.py]: remove

File: addons/sale_coupon/models/sale_order.py (hunk 15 above)
Comment (Feyensv) [reply addons/sale_coupon/models/sale_order.py]: --> #40336

File: addons/event_sale/models/sale_order.py [hunk 20]
```diff
@@ -33,6 +33,26 @@ class SaleOrderLine(models.Model):
...
+        # removing attendees
+        if not (self.event_id and self.event_ticket_id):
+            return super(SaleOrderLine, self).quantity_change()
+        registrations = self.env['event.registration'].search([
+                ('state', '!=', 'cancel'),
+                ('sale_order_id', '=', self.id),  # To avoid break on multi record set
+                ('event_ticket_id', '=', self.event_ticket_id.id),
+            ], order='create_date asc')
+        missing_qty = self.product_uom_qty - len(registrations)
+        # VFE TODO readonly when event soline
```
Comment (Feyensv) [inline addons/event_sale/models/sale_order.py]: (uom readonly)

File: addons/product/models/product_attribute.py [hunk 21]
```diff
@@ -388,11 +388,11 @@ class ProductTemplateAttributeValue(models.Model):
     attribute_line_id = fields.Many2one('product.template.attribute.line', required=True, ondelete='cascade', index=True)
 
     # configuration fields: the price_extra and the exclusion rules
-    price_extra = fields.Float(
+    price_extra = fields.Monetary(
```
Comment (Feyensv) [inline addons/product/models/product_attribute.py]: Revert to float ?

File: addons/event_sale/views/event_views.xml [hunk 22]
```diff
@@ -73,6 +73,8 @@
                                             <field name="name"/>
                                             <field name="product_id"/>
                                             <field name="price"/>
+                                            <field name="currency_id" invisible="1"/>
```
Comment (Feyensv) [inline addons/event_sale/views/event_views.xml]: add monetary widget on price.

File: addons/point_of_sale/static/src/js/models.js [hunk 23]
```diff
@@ -329,6 +329,8 @@ exports.PosModel = Backbone.Model.extend({
             if (self.config.use_pricelist) {
                 return [['id', 'in', self.config.available_pricelist_ids]];
             } else {
+                // VFE TODO if not use_pricelist do not load any pricelist ?
```
Comment (dbo-odoo) [inline addons/point_of_sale/static/src/js/models.js]: makeitso

File: addons/point_of_sale/static/src/js/models.js [hunk 24]
```diff
@@ -1382,6 +1392,12 @@ exports.Product = Backbone.Model.extend({
             }
 
             if (rule.base === 'pricelist') {
+                if (!rule.base_pricelist) {
+                    alert(_t(
+                        'An error occurred when loading product prices. ' +
+                        'Make sure all pricelists are available in the POS.'
```
Comment (dbo-odoo) [inline addons/point_of_sale/static/src/js/models.js]:
```
'Make sure all pricelists are available in the POS, '
'including any pricelist used as a base from another pricelist.'
```

File: addons/point_of_sale/static/src/js/models.js [hunk 25]
```diff
@@ -2837,7 +2853,11 @@ exports.Order = Backbone.Model.extend({
         return round_pr(this.orderlines.reduce((function(sum, orderLine) {
             sum += (orderLine.get_unit_price() * (orderLine.get_discount()/100) * orderLine.get_quantity());
             if (orderLine.display_discount_policy() === 'without_discount'){
+                // VFE TODO ref discount for point_of_sale also ?
```
Comment (dbo-odoo) [inline addons/point_of_sale/static/src/js/models.js]: check with pim & oma for pos questions

File: addons/product/models/product_pricelist.py [hunk 26]
```diff
@@ -119,32 +104,37 @@ def _compute_price_rule_get_items(self, products_qty_partner, date, uom_id, prod
...
+                'list_price',
+                uom=uom,
+                date=date,
+                currency=currency)
+            return {id: (prices[id], []) for id in products.ids}
+
+        self = self.with_company(self.company_id)
+        company = self.env.company  # pricelist company or current company
+        if not date:
+            date = fields.Date.today()
```
Comment (dbo-odoo) [inline addons/product/models/product_pricelist.py]: `date = fields.Date.context_today(products)`

File: addons/product/models/res_partner.py [hunk 27]
```diff
@@ -11,7 +11,7 @@ class Partner(models.Model):
...
-        inverse="_inverse_product_pricelist", company_dependent=False,
+        inverse="_inverse_product_pricelist",
```
Comment (dbo-odoo) [inline addons/product/models/res_partner.py]: can be left as a warning that something ~shitty~ fishy is going on

File: addons/product/models/product_template.py [hunk 28]
```diff
@@ -10,15 +10,96 @@
...
+    if not self:
+        return {}
+
+    self = self.with_company(company)
+
+    fixed_price, fixed_price_currency = False, None
+    if price_type == 'list_price' and self.env.context.get('fixed_sales_price') and self.env.context.get('fixed_sales_currency'):
+        self.ensure_one()
+        assert self.is_product_variant, _("A fixed price cannot be given for a template.")
+        # Fixed prices should only be given for a fixed product
```
[the remark this answers is not in the document]
Comment (Feyensv) [reply addons/product/models/product_template.py]: Fixed prices are only (at least for current implementations in Odoo) given after product_product specification.  But yes, we could leave it supported for templates also.

File: addons/product/models/product_template.py [hunk 29]
```diff
@@ -10,15 +10,96 @@
...
+    """
+    if not self:
+        return {}
+
+    self = self.with_company(company)
+
+    fixed_price, fixed_price_currency = False, None
+    if price_type == 'list_price' and self.env.context.get('fixed_sales_price') and self.env.context.get('fixed_sales_currency'):
+        self.ensure_one()
+        assert self.is_product_variant, _("A fixed price cannot be given for a template.")
```
[the remark this answers is not in the document]
Comment (Feyensv) [reply addons/product/models/product_template.py]: Yep, self.is_product_variant = True if self == product_product recordset

File: addons/product/models/product_template.py [hunk 30]
```diff
@@ -10,15 +10,96 @@
...
+            ptavs = ptavs - targeted_product.product_template_attribute_value_ids
+    ptav_extras = sum(ptavs.mapped('price_extra'))
+
+    if price_type == 'standard_price':
+        # standard_price field can only be seen by users in base.group_user
+        # Thus, in order to compute the sale price from the cost for users not in this group
+        # We fetch the standard price as the superuser
+        self = self.sudo()
+
+    if price_type == 'list_price':
```
[the remark this answers is not in the document]
Comment (Feyensv) [reply addons/product/models/product_template.py]: Currently, `list_price` is used for prices in Odoo, but I'm converting it to `lst_price` to make sure the extra prices are considered:
lst_price for product_template is related to list_price
lst_price for product_product = list_price + price_extra from ptavs.

File: addons/product/models/product_template.py [hunk 31]
```diff
@@ -10,15 +10,96 @@
...
+        date = fields.Date.today()
+
+    # VFE FIXME is this better for perf purposes ???
+    prices = dict.fromkeys(self.ids, 0.0)
+    for template in self:
+        # template = product.product or product.template
+        if not fixed_price:
+            # We should take lst_price if list_price :D
+            price = template[price_type] or 0.0
+            if ptav_extras and targeted_product_id and template.id == targeted_product_id:
```
[the remark this answers is not in the document]
Comment (Feyensv) [reply addons/product/models/product_template.py]: only a product, but i'll try to remove the p_id logic

File: addons/product/models/product_template.py [hunk 32]
```diff
@@ -183,47 +262,41 @@ def _compute_currency_id(self):
...
-    def _set_template_price(self):
-        if self._context.get('uom'):
-            for template in self:
-                value = self.env['uom.uom'].browse(self._context['uom'])._compute_price(template.price, template.uom_id)
-                template.write({'list_price': value})
-        else:
-            self.write({'list_price': self.price})
+    def _get_context_values(self):
+        pricelist_id = self.env.context.get('pricelist_id', False)
+        pricelist = self.env['product.pricelist'].browse(pricelist_id)
```
[the remark this answers is not in the document]
Comment (Feyensv) [reply addons/product/models/product_template.py]: product.pricelist(False) == product.pricelist()

browse(False) returns empty recordset :P

File: addons/product/models/res_partner.py (hunk 27 above)
Comment (Feyensv) [reply addons/product/models/res_partner.py]: The depends_context('company') is already there to say that, but ok ^^.

File: addons/event_sale/models/sale_order.py (hunk 20 above)
Comment (Feyensv) [reply addons/event_sale/models/sale_order.py]: Separate task defined.

File: addons/event_sale/security/event_security.xml [hunk 33]
```diff
@@ -5,4 +5,11 @@
             <field name="implied_ids" eval="[(4, ref('event.group_event_user'))]"/>
         </record>
     </data>
+
+    <record id="event_ticket_comp_rule" model="ir.rule">
+        <field name="name" >Event Ticket multi-company</field>
+        <field name="model_id" ref="model_event_event_ticket"/>
+        <field name="global" eval="True"/>
```
[the remark this answers is not in the document]
Comment (dbo-odoo) [reply addons/event_sale/security/event_security.xml]: for the record: global is a stored field but should obvsly not be set manually
also, it has a weird declaration ^^ https://github.com/odoo/odoo/blob/13.0/odoo/addons/base/models/ir_rule.py#L256

File: addons/point_of_sale/models/pos_config.py [hunk 34]
```diff
@@ -382,7 +385,7 @@ def _onchange_use_pricelist(self):
         using a pricelist for this iotbox.
         """
         if not self.use_pricelist:
-            self.pricelist_id = self._default_pricelist()
+            self.pricelist_id = None
```
Comment (Feyensv) [inline addons/point_of_sale/models/pos_config.py]: TODO merge the two onchanges

File: addons/point_of_sale/static/src/js/models.js [hunk 35]
```diff
@@ -2842,7 +2858,11 @@ exports.Order = Backbone.Model.extend({
...
+                // or do we consider that in this case, the pricelist is never based
+                // on cost or fixed prices on which the reduction should be based?
```
Comment (Feyensv) [inline addons/point_of_sale/static/src/js/models.js]: remove comments

File: addons/sale/models/sale.py [hunk 36]
```diff
@@ -375,7 +384,7 @@ def onchange_partner_id(self):
             values['user_id'] = user_id
 
         if self.env['ir.config_parameter'].sudo().get_param('account.use_invoice_terms') and self.env.company.invoice_terms:
-            values['note'] = self.with_context(lang=self.partner_id.lang).env.company.invoice_terms
+            values['note'] = self.env.company.invoice_terms
```
Comment (Feyensv) [inline addons/sale/models/sale.py]: Wrong

File: addons/sale_timesheet/models/project_sale_line_employee_map.py [hunk 37]
```diff
@@ -17,7 +17,8 @@ def _default_project_id(self):
     project_id = fields.Many2one('project.project', "Project", domain=[('billable_type', '!=', 'no')], required=True, default=_default_project_id)
     employee_id = fields.Many2one('hr.employee', "Employee", required=True)
     sale_line_id = fields.Many2one('sale.order.line', "Sale Order Item", domain=[('is_service', '=', True)], required=True)
-    price_unit = fields.Float(related='sale_line_id.price_unit', readonly=True)
+    price_unit = fields.Float(related='sale_line_id.price_unit')
```
Comment (Feyensv) [inline addons/sale_timesheet/models/project_sale_line_employee_map.py]: TODO display it with monetary widget when shown

File: addons/point_of_sale/models/pos_config.py [hunk 38]
```diff
@@ -109,17 +110,17 @@ def _compute_customer_html(self):
...
         domain=[('type', '=', 'sale')],
         help="Accounting journal used to post sales entries.",
-        default=_default_sale_journal)
+        default=_default_sale_journal, check_company=True)
     invoice_journal_id = fields.Many2one(
         'account.journal', string='Invoice Journal',
         domain=[('type', '=', 'sale')],
         help="Accounting journal used to create invoices.",
-        default=_default_invoice_journal)
+        default=_default_invoice_journal, check_company=True)
```
Comment (Feyensv) [inline addons/point_of_sale/models/pos_config.py]: to remove, already checked by a constraint.

**Comment by Feyensv** [comment]
On hold until further notice, some functional changes shall be merged in other PR's and tasks.

**Comment by legalsylvain** [comment]
what is the state of this PR ?

**Comment by Feyensv** [comment]
It could be resurrected in some months, but not soon :D.
It's a big mess and sthg should be done about it, but the change was quite big and nobody dared to merge it ^^.
Dunno when we'll give it another try, and it'll probably be done in smaller PR's to separate technical cleanup and functional changes.

**Comment by Feyensv** [comment]
Another PR cleaned most of the pricelist logic (https://github.com/odoo/odoo/pull/79605) and remaining changes are dropped or coming in other tasks/PRs.

**Pull request closed without merging**